Tournament tie break management

Molrn opened the pull request:
An improvement of the tie-breaks such that it allows the handling of all tie-break functions
- tie breaks are stored in the tournament table as a json list, with their options
- on migration, the tie breaks from the papi database are added to the event database

it lacks a UI (not my strong point gotta admit ^^') and translations / tie-break types probably gotta be tweaked a bit, but I think this is a solid basis to base the tie-break on

Molrn committed:
feat: Tie break management

Amaras commented on src/data/tie_break.py:
```diff
@@ -0,0 +1,476 @@
+            **self.options,
+        )
+
+    def _check_options(self):
+        for option, value in self.options.items():
+            if option not in self.type.options:
+                raise ValueError(
+                    f'Option "{option.value}" is not allowed '
+                    f'for tie break "{self.type.value}"'
+                )
+            if not isinstance(value, option.type):
+                raise ValueError(
+                    f'Unexpected value for option "{option.value}": '
+                    f'"{value}" (expected_type: {option.type})'
+                )
```
Some options are incompatible, I'll try to find the time to tell you which ones are.
In general, `papi_legacy` is incompatible with modifiers other than cuts

Molrn replied:
some incompatibilities could be removed by splitting the initial methods and having 2 separate ones. Example with sum of buchholz:
- from: sum_of_buchholz(fore_modifier, papi_legacy)
- to: sum_of_buchholz(papi_legacy), sum_of_fore_buchholz()

Amaras replied:
I don't think I quite like splitting the functions.
It would involve quite a lot of code repetition, which I am not overly fond of.
I'm open to the idea, but not in favour.

Molrn replied:
Could also be possible to have multiple TieBreakType entries for the same function:
- SUM_OF_BUCHHOLZ / SUM_OF_FORE_BUCHHOLZ  

To do so, we could have a `mandatory_options` a `forbidden_options` properties (SOB: `fore_modifier` forbidden, SOFB: `fore_modifier`=True mandatory / `papi_legacy` forbidden)
That could also be a way to split Buchholz into `Buchholz`, `Buchholz tronqué` and `Buchholz médian`

Molrn replied:
or even simpler: a `overriden_options` where options can't be selected:
- SOB: fore_modifier=False
- SOFB: fore_modifier=True, papi_legacy=False

pascalaubry replied:
@Molrn I let you mark the incompatible options.
Note: the options are OK for the Papi tie-breaks (tests OK).

Molrn replied:
Incompatible options:
- Buchholz: played_modifier=True / papi_legacy=True
- Sum of Buchholz: fore_modifier=True / papi_legacy=True  
- in general: papi_legacy=True / cut options (need to be 1, 2 or 3 depending on the number of rounds)  

All the other options appear to be compatible.  

Considering all the interaction issues appear to be related to papi_legacy, I suggest to treat papi_legacy differently than the other options. The purpose of this option seems to be to generate exactly the same tie-breaks as Papi, so generating tie-breaks with similar mistakes but more options than papi seems not relevant. Therefore, having in the form of each tie-break a checkbox 'papi_legacy' is in my opinion not a good idea.  

Instead, I suggest the following appraoch: having a global tie-break option 'papi_legacy'. If this option is selected, limit the amount of tie breaks to 3, and select the tie-breaks amongst the values of the PapiTieBreak Enum (similar to the current interface).  
As the options incompatibilites are not breaking anything or even generating false tie-breaks, I think it is safe not to implement this incompatibility in the TieBreak class, and to just rely on the interface to match these incompatibilities. It also saves the trouble of explaining them to the users.

Amaras reviewed (commented):
Not bad from what I have quickly read, I'll review more in depth later

Molrn committed:
adapt tie-break to papi storage

Molrn committed:
feat: kashdan tie-break

Molrn committed:
add papi legacy for kashdan tie-break

Molrn committed:
fix: compute tie-breaks after max_round instead of before

Molrn committed:
fix tournament form

pascalaubry commented on locale/fr/LC_MESSAGES/messages.po:
```diff
@@ -804,6 +804,134 @@ msgstr "échiquiers {board_numbers}"
+
+#: src/data/tie_break.py
+msgid "Perfect tournament performance"
+msgstr "Performance ELO parfaite"
```
I don't see what "**Perfect** tournament performance" stands for.
Shouldn't it be "Perfect tournament performance rating" (cf "Tournament performance rating")?

Molrn replied:
that's the name of the tie-break in the [FIDE handbook](https://handbook.fide.com/chapter/TieBreakRegulations082024#:~:text=10.3%C2%A0%C2%A0%20Perfect%20Tournament%20Performance%20(PTP)). As for the translations, I'm cross-referencing them right now with the 'Livre de l'Arbitre', which does not include all of them

pascalaubry commented on locale/fr/LC_MESSAGES/messages.po:
```diff
@@ -4645,6 +4773,15 @@ msgstr "Dernières rondes sans bye :"
+
+#: src/web/templates/admin/tournaments/tournament_modal.html
+#, python-format
+msgid "Tie-Break %(index)d:"
```
Enumerations are written like this on the other pages:
msgid "Tie-Break **#**%(index)d:"
msgstr "Départage **n°**%(index)d :" (space before ':' in French)

pascalaubry commented on src/data/tie_break.py:
```diff
@@ -0,0 +1,498 @@
+from collections.abc import Callable, Iterable
```
@Molrn It would probably be useful that you check the way Tornelo implemented the tie-breaks; for instance cut_top and cut_bottom are options of a single Buchholz tie-break.
![image](https://github.com/user-attachments/assets/1fc14d8e-5eb5-47f3-a6c7-c2da13adc7fd)
And maybe check we did not forget anything (but you probably checked the FIDE requirements):
![image](https://github.com/user-attachments/assets/358a9876-015d-44b1-98d1-85da458769b8)

Molrn replied:
in  #114 @Amaras implemented all the individual tie-breaks described in the [FIDE Handbook section C07](https://handbook.fide.com/chapter/TieBreakRegulations082024), which are also the ones used in this TEC tie-break exercises documents.  

I also added Kashdan in this PR in order to have all the Tie-breaks defined in Papi. I think with FIDE tie-breaks + Papi, we are good. After that point, I suggest to only add tie-breaks if they are requested. If there are other tie-breaks that are relevant to you don't hesitate (for example Arranz in Tornelo that I did not know about) ;)  

Besides Arranz, the other tie-break Tornelo has that we haven't explicitely defined is Start number. It is used in Papi-web and [Tornelo](https://tornelo.com/knowledge-base/how-to-setup-tie-breaks/#:~:text=This%20tie%2Dbreak%20is%20always%20used%20as%20a%20last%20resort%20even%20if%20not%20chosen%20in%20the%20settings.) as default tie-break, but not explicitely in Papi-web. I we don't explicitely include it like Tornelo, I think we would need to at some point mention it in either documentation or directly in the UI

Molrn replied:
Also about default tie-break: I still have no idea what Papi uses as default tie-break, and I ran out of ideas of what it could be. We probably will need to figure this out at some point  

it is not:
- the performance (even if it is implied in the 'livre de l'arbitre' section 4.2.8)
- any other tie-break method
- the starting rank
- the pairing number (tied to the starting rank)
- the id (because why not)
- Alphabetical (might be getting desperate here)

pascalaubry replied:
+1: at this time only the tie-breaks requested by the FIDE should be implemented.

pascalaubry commented on src/database/sql/create_event.sql:
```diff
@@ -85,6 +85,7 @@ CREATE TABLE `tournament` (
     `last_ffe_upload` FLOAT NOT NULL DEFAULT 0.0,
     `last_ffe_rules_upload` FLOAT NOT NULL DEFAULT 0.0,
     `last_chessevent_download_md5` TEXT,
+    `tie_breaks` TEXT NOT NULL DEFAULT '[]',
```
Optional fields are generally left NULL when empty, this makes possible the evolution of the storage format with no change to the database structure (see info.timer_colors|timer_delays and timer.colors|delays).

pascalaubry commented on src/web/controllers/admin/tournament_admin_controller.py:
```diff
@@ -196,6 +198,25 @@ def _admin_validate_tournament_update_data(
+                and tournament.stored_tournament.filename == filename
+                and tournament.file_exists
+            ):
+                tournament.update_tie_breaks(
```
tournament.update_tie_breaks() should be called in _admin_tournament_update(), not here (in this method we only validate the data without changing it).

pascalaubry commented on src/web/templates/admin/tournaments/tournament_modal.html:
```diff
@@ -285,6 +285,40 @@ <h4>
                     </div>
+                    {% if edit_tie_breaks %}
+                        <h4>
+                            {{ _('Tie-Breaks') }}
```
Tie-breaks (no upper-case B)

pascalaubry reviewed (changes requested):
Everything clear to me except the place of the call to tournament.update_tie_breaks().
Needs another review on the tie-break code itself anyway.

Molrn committed:
fixes

- DB field tie_breaks: default null
- move update method to _admin_tournament_update
- translations

pascalaubry commented:
Please fill 85-database.md for the new fields.

Molrn committed:
Doc: update database schema

Molrn committed:
fix tie-break warnings

pascalaubry commented on docs/85-database.md:
```diff
@@ -83,7 +77,9 @@
-| `chessevent_id`                       | `INTEGER` | REFERENCES `chessevent`(`id`)              | L'identifiant de la connexion à ChessEvent.                                                                                                                                             |
+| `chessevent_user_id`                  | `TEXT`    |                                            | L'identifiant utilisé pour le connexion à la plateforme ChessEvent.                                                                                                                     |
+| `chessevent_password`                 | `TEXT`    |                                            | Le mot de passe utilisé pour le connexion à la plateforme ChessEvent.                                                                                                                   |
+| `chessevent_event_id`                 | `TEXT`    |                                            | L'identifiant de l'évènement ChessEvent sur la plateforme.                                                                                                                              |
```
These changes had been forgotten, thank you.

pascalaubry commented on docs/85-database.md:
```diff
@@ -95,6 +91,10 @@
 | `last_chessevent_download_md5`        | `TEXT`    |                                            | L'empreinte du dernier téléchargement du tournoi depuis la plateforme ChessEvent.                                                                                                       |
+| `first_board_number`                  | `INTEGER` |                                            | Le numéro du premier échiquier.                                                                                                                                                         |
+| `paired_bye_points`                   | `FLOAT`   |                                            | Points attribués aux exempts.                                                                                                                                                           |
+| `max_byes`                            | `INTEGER` |                                            | Le nombre maximum de byes auquel un·e joueur·euse peut prétendre                                                                                                                        |
```
These too, btw can you add last_rounds_no_byes?

Molrn committed:
add last_rounds_no_byes to doc
